src/modules/interface/controller_indi.h

```diff
@@ -0,0 +1,125 @@
+/**
+ *    ||          ____  _ __
+ * +------+      / __ )(_) /_______________ _____  ___
+ * | 0xBC |     / __  / / __/ ___/ ___/ __ `/_  / / _ \
+ * +------+    / /_/ / / /_/ /__/ /  / /_/ / / /_/  __/
+ *  ||  ||    /_____/_/\__/\___/_/   \__,_/ /___/\___/
+ *
+ * Crazyflie control firmware
+ *
+ * Copyright (C) 2011-2016 Bitcraze AB
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, in version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * controller_indi.h - INDI Controller Interface
+ */
+#ifndef __CONTROLLER_INDI_H__
+#define __CONTROLLER_INDI_H__
+
+#include "stabilizer_types.h"
+#include "filter.h"
+#include "math3d.h"
+
+#define PI 3.14159265358979323846f
+
+//The G values are scaled to avoid numerical problems during the estimation
+#define INDI_EST_SCALE 0.001f
+
+// these parameters are used in the filtering of the angular acceleration
+#define STABILIZATION_INDI_FILT_CUTOFF 8.0f
+
+// the yaw sometimes requires more filtering
+#define STABILIZATION_INDI_FILT_CUTOFF_R STABILIZATION_INDI_FILT_CUTOFF
+
+#define STABILIZATION_INDI_ESTIMATION_FILT_CUTOFF 4.0f
+
+#define PERIODIC_FREQUENCY 512.0f
+
+// these parameters are used in the filtering of the angular acceleration
+#define STABILIZATION_INDI_MAX_RATE 6.0f
+#define STABILIZATION_INDI_MAX_R 120.0f
+#define STABILIZATION_INDI_G1_P 0.0639f
+#define STABILIZATION_INDI_G1_Q 0.0361f
+#define STABILIZATION_INDI_G1_R 0.0022f
+#define STABILIZATION_INDI_G2_R 0.1450f
+#define STABILIZATION_INDI_REF_ERR_P 600.0f
+#define STABILIZATION_INDI_REF_ERR_Q 600.0f
+#define STABILIZATION_INDI_REF_ERR_R 600.0f
+#define STABILIZATION_INDI_REF_RATE_P 28.0f
+#define STABILIZATION_INDI_REF_RATE_Q 28.0f
+#define STABILIZATION_INDI_REF_RATE_R 28.0f
```

Comment:
These parameters are surely too aggressive for the Crazyflie, try to start with 100.0f and 14.0f for ERR and RATE respectively

Reply:
Done! I still do not have this sensibility, I lack symbol grounding in the real world for these values. I think as soon as we start experiment I will have a better notion.

Reply:
STABILIZATION_INDI_REF_RATE_P this one is the amount of angular acceleration in rad/s2 that you want per rad/s rate error. The other one is actually based on a quaternion error. Here is the quaternion for a certain rotation theta:
![image](https://user-images.githubusercontent.com/2464596/62368199-3bebac00-b52c-11e9-9263-10cdfe23ef2b.png)
Which will give a value in u_x etc that is approx. half as big. So the other gain, STABILIZATION_INDI_REF_ERR_R, should be halved if we feed back the angles instead. But then it is [gain] (rad/s ) / rad of attitude error.

Actually, for this latter one, the implementation matters:
1) ang_acc_desired = P*att_err - D* rate
2) ang_acc_desired = D*(P*att_err - rate)
That gain of 100 that I mentioned was actually based on method 1, with quaternion feedback. If we use method 2 with angle feedback, we have to do 100/D/2 = 100/14/2 = 3.57

For first order actuator dynamics with tau = 1/16, from experience and looking at the poles of the total transfer function I know that these gains should perform well. Also consider section 2.1 in "Cascaded Incremental Nonlinear Dynamic Inversion Control for MAV Disturbance Rejection"

Sorry for the lengthy explanation and giving you wrong numbers at first.

Reply:
So in the end we should be using 3.57f for ERR and 14.0f for RATE respectively (before calculating with the experiments)?

Reply:
Yes, if angles are in radians